examples/streamlit/src/run_app.py

```diff
@@ -0,0 +1,36 @@
+import os
+import subprocess
+import sys
+import webbrowser
+
+from src.config import EnvironmentalVariableNames as EnvVar, get_env
+
+
+def main():
+
+    # Getting path to python executable (full path of deployed python on Windows)
+    executable = sys.executable
+
+    # Open browser tab. May temporarily display error until streamlit server is started.
+    command = 'import time; import webbrowser; time.sleep(3); webbrowser.open("http://localhost:8501")'
+    subprocess.run(f"{executable} -c '{command}'", shell=True)
+
+    # Run streamlit server
+    path_to_main = os.path.join(
+        get_env(EnvVar.WORKING_DIR), "src", "app.py"
+    )
+    result = subprocess.run(
+        f"{executable} -m streamlit run {path_to_main} --server.headless=true --global.developmentMode=false",
+        shell=True,
+        capture_output=True,
+        text=True,
+    )
+
+    # These are printed only when server is stopped.
+    # NOTE: you have to manually stop streamlit server killing process.
+    print(result.stdout)
+    print(result.stderr)
```

Comment:
This simple option is good for subprocesses that finish quickly, but it isn't ideal if you have it running for a long time, because the output will build up in memory.

One simple option is just to discard output:

```python
from subprocess import run, DEVNULL

run(..., stdout=DEVNULL, stderr=DEVNULL)
```

If you want to capture the output and print it as it comes, that's possible, but a bit more complicated:

```python
from subprocess import Popen, PIPE, STDOUT

proc = Popen(..., stdin=PIPE, stdout=PIPE, stderr=STDOUT, text=True)
proc.stdin.close()

while True:
    s = proc.stdout.read()
    if not s:
        break
    print(s, end='')

proc.wait()
```

Reply:
Thanks for this snippet, I chaged the code using this since it seems more general to me. Unfortunately, this does not work as expected, that is the output is not written as it comes. I checked the `C:\Users{username}\AppData\Roaming` where the app is supposed to write, and it remains blank as far as the streamlit process is stopped. Then it shows the output (the standard Streamlit welcome message with the ip address to reach the app)

Reply:
That probably means that the output is being buffered, either going from your launcher to the file, or from streamlit to your launcher. Python saves up writes for a while and then does one bigger write, which is more efficient than lots of small writes. But the efficiency difference shouldn't matter for a fairly small amount of output.

If you want to make it appear in the file as it comes, try adding `flush=True` to the `print()` call, and if that doesn't do it, add the [`-u` option](https://docs.python.org/3/using/cmdline.html#cmdoption-u) before `-m` where you launch the subprocess. But so long as I'm right about the buffering, it doesn't matter too much - all the output should be there when the program exits, and it should avoid anything getting stuck if streamlit prints e.g. logging output.

Reply:
I understand. I tested both your suggestions, but none of them worked. The output still gets printed once the streamlit server has exited.

This is not really a problem, since the important thing is that there is way to check the output of the app if something goes wrong. From your above message I understood that there was the possibility to get the logs in "real time", but it was my misunderstanding. I think we can leave it as is, it fits for the purpose.

Reply:
Fair enough, thanks for testing it! :slightly_smiling_face: